File: api/routes/metrics.py
"""Metrics and evaluation endpoints.

Routes:
    POST  /api/metrics/evaluate        — Evaluate a single image with one model
    POST  /api/metrics/compare         — Run one image through multiple models
    POST  /api/metrics/batch_evaluate  — Evaluate all test samples for a model
"""

from flask import Blueprint, request, jsonify, current_app
from api.services.metrics_service import MetricsService

bp = Blueprint('metrics', __name__)
_svc = MetricsService()


@bp.route('/evaluate', methods=['POST'])
def evaluate_image():
    """Compute PSNR, SSIM, LPIPS for a single prediction.

    JSON body:
        {
            "image_path": str,    # path to source image
            "model": str,
            "checkpoint": str
        }
    """
    data = request.get_json(force=True) or {}
    try:
        result = _svc.evaluate_single(
            data['image_path'],
            data.get('model', 'unet'),
            data.get('checkpoint', ''),
        )
    except Exception as exc:
        return jsonify({'error': str(exc)}), 500
    return jsonify(result)


@bp.route('/compare', methods=['POST'])
def compare_models():
    """Run one image through multiple model configurations and return metrics.

    JSON body:
        {
            "image_path": str,
            "models": [
                {"model": str, "checkpoint": str, "label": str},
                ...
            ]
        }
    """
    data = request.get_json(force=True) or {}
    try:
        result = _svc.compare_models(
            data['image_path'],
            data.get('models', []),
        )
    except Exception as exc:
        return jsonify({'error': str(exc)}), 500
    return jsonify(result)


@bp.route('/batch_evaluate', methods=['POST'])
def batch_evaluate():
    """Evaluate all images in data/test_samples/ for a given model.

    JSON body:
        {
            "model": str,
            "checkpoint": str
        }
    """
    data = request.get_json(force=True) or {}
    try:
        result = _svc.evaluate_samples(
            data.get('model', 'unet'),
            data.get('checkpoint', ''),
            sample_dir=current_app.config.get('SAMPLE_DIR', 'data/test_samples'),
        )
    except Exception as exc:
        return jsonify({'error': str(exc)}), 500
    return jsonify(result)

File: frontend/src/api/metrics.ts
import api from './client'
import type { ModelType, EvalResult, ColorizeResult, BenchmarkModelConfig, BenchmarkResult } from '@/types'

export const metricsApi = {
  evaluateSingle: (imagePath: string, model: ModelType, checkpoint: string) =>
    api.post<{ model: string; checkpoint: string; metrics: { psnr: number | null; ssim: number | null } }>(
      '/metrics/evaluate',
      { image_path: imagePath, model, checkpoint },
    ).then((r) => r.data),

  batchEvaluate: (model: ModelType, checkpoint: string) =>
    api.post<EvalResult>(
      '/metrics/batch_evaluate',
      { model, checkpoint },
      { timeout: 180_000 },
    ).then((r) => r.data),

  compareModels: (
    imagePath: string,
    models: { model: ModelType; checkpoint: string; label: string }[],
  ) =>
    api.post<(ColorizeResult & { label: string })[]>(
      '/metrics/compare',
      { image_path: imagePath, models },
    ).then((r) => r.data),

  benchmark: (models: BenchmarkModelConfig[], maxImages?: number, imageDir?: string) =>
    api.post<BenchmarkResult>(
      '/metrics/benchmark',
      { models, max_images: maxImages, image_dir: imageDir || undefined },
      { timeout: 600_000 },
    ).then((r) => r.data),
}
